src/protojson/JobErrorMsg.cc

```diff
@@ -0,0 +1,111 @@
+#include "protojson/JobErrorMsg.h"
+
+#include <stdexcept>
+
+// Qserv headers
+#include "http/Client.h"
+#include "http/MetaModule.h"
+#include "http/RequestBodyJSON.h"
+#include "util/common.h"
+#include "util/TimeUtils.h"
+
+// LSST headers
+#include "lsst/log/Log.h"
+
+using namespace std;
+using namespace nlohmann;
+
+namespace {
+LOG_LOGGER _log = LOG_GET("lsst.qserv.protojson.JobErrorMsg");
+}  // namespace
+
+namespace lsst::qserv::protojson {
+
+JobErrorMsg::Ptr JobErrorMsg::create(std::string const& replicationInstanceId,
+                                     std::string const& replicationAuthKey, std::string const& workerIdStr,
+                                     std::string const& czarName, CzarIdType czarId, QueryId queryId,
+                                     UberJobId uberJobId, int errorCode, std::string const& errorMsg) {
+    auto jrMsg = Ptr(new JobErrorMsg(replicationInstanceId, replicationAuthKey));
+    jrMsg->_workerId = workerIdStr;
+    jrMsg->_czarName = czarName;
+    jrMsg->_czarId = czarId;
+    jrMsg->_queryId = queryId;
+    jrMsg->_uberJobId = uberJobId;
+    jrMsg->_errorCode = errorCode;
+    jrMsg->_errorMsg = errorMsg;
+    return jrMsg;
+}
+
+JobErrorMsg::Ptr JobErrorMsg::createFromJson(nlohmann::json const& jsWReq,
+                                             std::string const& replicationInstanceId,
+                                             std::string const& replicationAuthKey) {
+    string const fName("JobErrorMsg::createFromJson");
+    LOGS(_log, LOG_LVL_DEBUG, fName);
+    try {
+        if (jsWReq["version"] != http::MetaModule::version) {
+            LOGS(_log, LOG_LVL_ERROR, fName << " bad version");
+            return nullptr;
+        }
+
+        // Presumably, if these were wrong, it wouldn't have gotten this far.
+        auto repliInstId = http::RequestBodyJSON::required<string>(jsWReq, "instance_id");
+        auto repliAuthKey = http::RequestBodyJSON::required<string>(jsWReq, "auth_key");
+
+        auto jrMsg = create(repliInstId, repliAuthKey);
+        jrMsg->_workerId = http::RequestBodyJSON::required<string>(jsWReq, "workerid");
+        jrMsg->_czarName = http::RequestBodyJSON::required<string>(jsWReq, "czar");
+        jrMsg->_czarId = http::RequestBodyJSON::required<CzarIdType>(jsWReq, "czarid");
+        jrMsg->_queryId = http::RequestBodyJSON::required<QueryId>(jsWReq, "queryid");
+        jrMsg->_uberJobId = http::RequestBodyJSON::required<UberJobId>(jsWReq, "uberjobid");
+        jrMsg->_errorMsg = http::RequestBodyJSON::required<string>(jsWReq, "errorMsg");
+        jrMsg->_errorCode = http::RequestBodyJSON::required<int>(jsWReq, "errorCode");
+        return jrMsg;
+    } catch (invalid_argument const& exc) {
+        LOGS(_log, LOG_LVL_ERROR, string("JobErrorMsg::createJson invalid ") << exc.what());
+    }
+    return nullptr;
+}
+
+json JobErrorMsg::serializeJson() {
+    shared_ptr<json> jsJrReqPtr = make_shared<json>();
+    json& jsJr = *jsJrReqPtr;
```

Comment:
Same strange pattern of creating the pointer.

Do you have it in the rest of the new Qserv code? If so then please get rid of this pattern. Pointers should be avoided. There are very little benefits of using those in Qserv.